Micronaut 4.2.0 breaks grpc services

### Expected Behavior

GRPC serialization back to a client reply should work.

**NOTE: this does not happen in the projects below because the 3rd party patch was applied to them.  This allows the grpc services to work in micronaut 4.2.0. **



### Actual Behaviour

GRPC throws exceptions

## Examples

I have a number of simple grpc projects in my repo that run micronaut services.  Once I bumped the service to 4.2.0, the OOTB functionality no longer works for grpc blocking stubs and regular stubs.  Below are the URLs to the projects (there's one dependent project as well, which matches the grpc running version of 1.59.0.

The sample applications where this depends on grpc servers are as follows:
https://github.com/krickert/search_indexer/tree/master/services/grpc-services/grpc-nlp-service
https://github.com/krickert/search_indexer/tree/master/services/grpc-services/grpc-vectorizer-service
https://github.com/krickert/search_indexer/tree/master/services/pipeline-processor

The tests that break are any test that uses the standard grpc transport.

## Barebones test (follows the same as the guide docs)

Here is the most barebones test:

https://github.com/krickert/grpc_micronaut_420_broken_example

This uses jdk17 (amazon latest), maven 3.9.5 on the latest ubuntu desktop.  Also repeatable on jdk17 on osx.  Didn't test anywhere else yet since I see consistent results.

## output from test

```
[ERROR] com.broken.grpc.GrpcExampleTest.testServerEndpoint -- Time elapsed: 0.165 s <<< ERROR!
io.grpc.StatusRuntimeException: UNKNOWN
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:268)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:249)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:167)
	at com.broken.grpc.GrpcExampleServiceGrpc$GrpcExampleServiceBlockingStub.send(GrpcExampleServiceGrpc.java:160)
	at com.broken.grpc.GrpcExampleTest.testServerEndpoint(GrpcExampleTest.java:79)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:157)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:114)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.UnsupportedOperationException
	at io.grpc.netty.AbstractHttp2Headers.isEmpty(AbstractHttp2Headers.java:40)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:419)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:352)
	at io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onHeadersRead(Http2InboundFrameLogger.java:56)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:476)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:484)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:253)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:159)
	at io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:188)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:393)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:453)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
```
The above is returned for the non-blocking test... 

The following is returned for the blocking test:
```
11:19:53.868 [io-executor-thread-2] ERROR com.broken.grpc.GrpcExampleTest - Not implemented
io.grpc.StatusRuntimeException: UNKNOWN
	at io.grpc.Status.asRuntimeException(Status.java:537)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:574)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:72)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:742)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.UnsupportedOperationException: null
	at io.grpc.netty.AbstractHttp2Headers.isEmpty(AbstractHttp2Headers.java:40)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:419)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:352)
	at io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onHeadersRead(Http2InboundFrameLogger.java:56)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:476)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:484)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:253)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:159)
	at io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:188)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:393)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:453)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
```

In both instances, the service is called, it seems the serialization is the broken part.




### Steps To Reproduce

1. Linux ubuntu
2. jdk17 / maven 3.9.5
3. `git clone https://github.com/krickert/grpc_micronaut_420_broken_example.git`
4. `cd grpc_micronaut_420_broken_example`
5. `mvn test`

### Environment Information

Latest ubuntu
jdk 17 
```
~/IdeaProjects/grpc_micronaut_420_broken_example$ java -version
openjdk version "17.0.9" 2023-10-17 LTS
OpenJDK Runtime Environment Corretto-17.0.9.8.1 (build 17.0.9+8-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.9.8.1 (build 17.0.9+8-LTS, mixed mode, sharing)
```


### Example Application

https://github.com/krickert/grpc_micronaut_420_broken_example

### Version

4.2.0

--- a/buildSrc/build.gradle
+++ b/buildSrc/build.gradle
@@ -9,6 +9,6 @@ repositories {
 
 dependencies {
     implementation libs.gradle.micronaut
-    implementation (libs.kotlin.gradle.plugin)
-    implementation (libs.kotlin.gradle.allopen)    
+    implementation libs.kotlin.gradle.plugin
+    implementation libs.kotlin.gradle.allopen
 }

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,6 +1,5 @@
 projectVersion=4.2.0-SNAPSHOT
 projectGroup=io.micronaut.grpc
-protobufGradleVersion=0.9.4
 title=Micronaut gRPC
 projectDesc=Integration between Micronaut and gRPC
 projectUrl=https://micronaut.io

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,8 +1,6 @@
 projectVersion=4.2.0-SNAPSHOT
 projectGroup=io.micronaut.grpc
 protobufGradleVersion=0.9.4
-protobufJavaVersion=3.25.0
-grpcVersion=1.59.0
 title=Micronaut gRPC
 projectDesc=Integration between Micronaut and gRPC
 projectUrl=https://micronaut.io

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -1,35 +1,44 @@
 [versions]
-managed-grpc = '1.59.0'
-managed-protobuf = '3.25.0'
+managed-grpc = '1.59.1'
+managed-protobuf = '3.25.1'
 managed-protobuf-gradle = '0.9.4'
-managed-grpc-kotlin = '1.4.0'
+managed-grpc-kotlin = '1.4.1'
 
 jackson-datatype-protobuf = '0.9.14'
 jaeger-core = '1.8.1'
 opentracing-grpc = '0.2.3'
 opentracing-mock = '0.33.0'
+graal-svm = '23.1.1'
+groovy = '4.0.15'
+javax-annotation-api = '1.3.2'
+kotlin = '1.9.20'
+spock = '2.3-groovy-4.0'
 
-micronaut = "4.1.11"
-micronaut-platform = "4.0.6"
-micronaut-docs = "2.0.0"
-micronaut-test = "4.0.0"
-micronaut-gradle-plugin = "4.1.2"
-
-graal-svm = "23.1.1"
-groovy = "4.0.15"
-javax-annotation-api = "1.3.2"
-kotlin = "1.9.20"
-spock = "2.3-groovy-4.0"
-
-micronaut-discovery = "4.0.1"
-micronaut-reactor = "3.1.0"
-micronaut-tracing = "6.2.0"
-micronaut-tracing-legacy = "3.2.7"
-micronaut-validation = "4.2.0"
+micronaut = '4.1.11'
+micronaut-platform = '4.1.6'
+micronaut-docs = '2.0.0'
+micronaut-gradle-plugin = '4.2.0'
+micronaut-discovery-client = '4.1.0'
+micronaut-reactor = '3.1.0'
+micronaut-tracing = '6.2.0'
+micronaut-tracing-legacy = '3.2.7'
+micronaut-validation = '4.2.0'
+micronaut-kotlin = '4.1.0'
+micronaut-logging = '1.1.2'
+micronaut-test = '4.0.2'
 
 [libraries]
 # Core
 micronaut-core = { module = 'io.micronaut:micronaut-core-bom', version.ref = 'micronaut' }
+micronaut-docs-asciidoc-config-props = { module = 'io.micronaut.docs:micronaut-docs-asciidoc-config-props', version.ref = 'micronaut-docs' }
+micronaut-tracing-legacy = { module = 'io.micronaut:micronaut-tracing', version.ref = 'micronaut-tracing-legacy' }
+micronaut-reactor = { module = 'io.micronaut.reactor:micronaut-reactor-bom', version.ref = 'micronaut-reactor' }
+micronaut-tracing = { module = 'io.micronaut.tracing:micronaut-tracing-bom', version.ref = 'micronaut-tracing' }
+micronaut-discovery-client = { module = "io.micronaut.discovery:micronaut-discovery-client", version.ref = "micronaut-discovery-client" }
+micronaut-validation = { module = 'io.micronaut.validation:micronaut-validation-bom', version.ref = 'micronaut-validation' }
+micronaut-kotlin = { module = 'io.micronaut.kotlin:micronaut-kotlin-bom', version.ref = 'micronaut-kotlin' }
+micronaut-logging = { module = 'io.micronaut.logging:micronaut-logging-bom', version.ref = 'micronaut-logging' }
+micronaut-test = { module = 'io.micronaut.test:micronaut-test-bom', version.ref = 'micronaut-test' }
 
 managed-grpc-protobuf = { module = 'io.grpc:grpc-protobuf', version.ref = 'managed-grpc' }
 managed-grpc-stub = { module = 'io.grpc:grpc-stub', version.ref = 'managed-grpc' }
@@ -38,19 +47,18 @@ managed-protoc-grpc = { module = 'io.grpc:protoc-gen-grpc-java', version.ref = '
 managed-grpc-kotlin-stub = { module = 'io.grpc:grpc-kotlin-stub', version.ref = 'managed-grpc-kotlin' }
 
 managed-protobuf-java = { module = 'com.google.protobuf:protobuf-java', version.ref = 'managed-protobuf' }
+managed-protobuf-javalite = { module = 'com.google.protobuf:protobuf-javalite', version.ref = 'managed-protobuf' }
+managed-protobuf-kotlin = { module = 'com.google.protobuf:protobuf-kotlin', version.ref = 'managed-protobuf' }
+managed-protobuf-kotlin-lite = { module = 'com.google.protobuf:protobuf-kotlin-lite', version.ref = 'managed-protobuf' }
 managed-protobuf-java-util = { module = 'com.google.protobuf:protobuf-java-util', version.ref = 'managed-protobuf' }
 managed-protoc = { module = 'com.google.protobuf:protoc', version.ref = 'managed-protobuf' }
 
-micronaut-discovery-client = { module = "io.micronaut.discovery:micronaut-discovery-client", version.ref = "micronaut-discovery" }
-micronaut-docs-asciidoc-config-props = { module = 'io.micronaut.docs:micronaut-docs-asciidoc-config-props', version.ref = 'micronaut-docs' }
-micronaut-tracing-legacy = { module = "io.micronaut:micronaut-tracing", version.ref = "micronaut-tracing-legacy" }
-micronaut-validation = { module = "io.micronaut.validation:micronaut-validation-bom", version.ref = "micronaut-validation" }
-
-graal-svm = { module = "org.graalvm.nativeimage:svm", version.ref = "graal-svm" }
+graal-svm = { module = 'org.graalvm.nativeimage:svm', version.ref = 'graal-svm' }
 grpc-netty = { module = 'io.grpc:grpc-netty' }
 
 jackson-datatype-protobuf = { module = 'com.hubspot.jackson:jackson-datatype-protobuf', version.ref = 'jackson-datatype-protobuf' }
-javax-annotation-api = { module = "javax.annotation:javax.annotation-api", version.ref = "javax-annotation-api" }
+javax-annotation-api = { module = 'javax.annotation:javax.annotation-api', version.ref = 'javax-annotation-api' }
+groovy = {module = 'org.apache.groovy:groovy' }
 
 jaeger-core = { module = 'io.jaegertracing:jaeger-core', version.ref = 'jaeger-core' }
 
@@ -60,14 +68,16 @@ netty-tcnative-boringssl-static = { module = 'io.netty:netty-tcnative-boringssl-
 opentracing-grpc = { module = 'io.opentracing.contrib:opentracing-grpc', version.ref = 'opentracing-grpc' }
 opentracing-mock = { module = 'io.opentracing:opentracing-mock', version.ref = 'opentracing-mock' }
 
-## BOMs
-micronaut-reactor = { module = 'io.micronaut.reactor:micronaut-reactor-bom', version.ref = "micronaut-reactor" }
-micronaut-tracing = { module = "io.micronaut.tracing:micronaut-tracing-bom", version.ref = "micronaut-tracing" }
+jackson-module-kotlin = { module = 'com.fasterxml.jackson.module:jackson-module-kotlin' }
+kotlin-stdlib-jdk8 = { module = 'org.jetbrains.kotlin:kotlin-stdlib-jdk8' }
+kotlin-reflect = { module = 'org.jetbrains.kotlin:kotlin-reflect' }
+kotlinx-coroutines-core = { module = 'org.jetbrains.kotlinx:kotlinx-coroutines-core' }
 
+## BOMs
 boms-protobuf = { module = 'com.google.protobuf:protobuf-bom', version.ref = 'managed-protobuf' }
 boms-grpc = { module = 'io.grpc:grpc-bom', version.ref = 'managed-grpc' }
 
-gradle-micronaut = { module = "io.micronaut.gradle:micronaut-gradle-plugin", version.ref = "micronaut-gradle-plugin" }
+gradle-micronaut = { module = 'io.micronaut.gradle:micronaut-gradle-plugin', version.ref = 'micronaut-gradle-plugin' }
 kotlin-gradle-plugin = { module = 'org.jetbrains.kotlin:kotlin-gradle-plugin', version.ref = 'kotlin' }
 kotlin-gradle-allopen = { module = 'org.jetbrains.kotlin:kotlin-allopen', version.ref = 'kotlin' }
 

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -11,11 +11,11 @@ opentracing-mock = '0.33.0'
 graal-svm = '23.1.1'
 groovy = '4.0.15'
 javax-annotation-api = '1.3.2'
-kotlin = '1.9.20'
+kotlin = '1.9.21'
 spock = '2.3-groovy-4.0'
 
-micronaut = '4.1.11'
-micronaut-platform = '4.1.6'
+micronaut = '4.2.0'
+micronaut-platform = '4.2.0'
 micronaut-docs = '2.0.0'
 micronaut-gradle-plugin = '4.2.0'
 micronaut-discovery-client = '4.1.0'
@@ -77,10 +77,10 @@ kotlinx-coroutines-core = { module = 'org.jetbrains.kotlinx:kotlinx-coroutines-c
 boms-protobuf = { module = 'com.google.protobuf:protobuf-bom', version.ref = 'managed-protobuf' }
 boms-grpc = { module = 'io.grpc:grpc-bom', version.ref = 'managed-grpc' }
 
+## plugins
 gradle-micronaut = { module = 'io.micronaut.gradle:micronaut-gradle-plugin', version.ref = 'micronaut-gradle-plugin' }
 kotlin-gradle-plugin = { module = 'org.jetbrains.kotlin:kotlin-gradle-plugin', version.ref = 'kotlin' }
 kotlin-gradle-allopen = { module = 'org.jetbrains.kotlin:kotlin-allopen', version.ref = 'kotlin' }
 
 [plugins]
 protobuf = { id = 'com.google.protobuf', version.ref = 'managed-protobuf-gradle' }
-

--- a/grpc-client-runtime/build.gradle
+++ b/grpc-client-runtime/build.gradle
@@ -10,7 +10,7 @@ dependencies {
     api libs.managed.grpc.protobuf
     api libs.managed.grpc.stub
 
-    compileOnly libs.micronaut.discovery.client
+    compileOnly mn.micronaut.discovery.core
 
     implementation libs.grpc.netty
     implementation mn.reactor
@@ -19,7 +19,8 @@ dependencies {
     testCompileOnly libs.javax.annotation.api
 
     testImplementation projects.micronautGrpcServerRuntime
-    testImplementation libs.micronaut.discovery.client
+    testImplementation mn.micronaut.discovery.core
+    testImplementation mnReactor.micronaut.reactor.http.client
     testImplementation mn.micronaut.jackson.databind
     testImplementation libs.opentracing.mock
 

--- a/test-suite-groovy/build.gradle
+++ b/test-suite-groovy/build.gradle
@@ -1,6 +1,6 @@
 plugins {
-    id 'groovy'
-    id "com.google.protobuf" version "$protobufGradleVersion"
+    id "groovy"
+    alias(libs.plugins.protobuf)
     id "io.micronaut.minimal.application"
     id "io.micronaut.build.internal.grpc-base"
 }

--- a/test-suite-groovy/build.gradle
+++ b/test-suite-groovy/build.gradle
@@ -5,6 +5,13 @@ plugins {
     id "io.micronaut.build.internal.grpc-base"
 }
 
+// tag::variables[]
+ext {
+    grpcVersion = libs.versions.managed.grpc.asProvider().get()
+    protobufVersion = libs.versions.managed.protobuf.asProvider().get()
+}
+// end::variables[]
+
 mainClassName = "helloworld.Application"
 micronaut {
     version.set(libs.versions.micronaut.platform.get())
@@ -18,25 +25,26 @@ micronaut {
 }
 
 dependencies {
-    compileOnly "io.grpc:grpc-stub"
-    compileOnly "javax.annotation:javax.annotation-api"
-    compileOnly "io.micronaut.validation:micronaut-validation-processor"
-    implementation "io.micronaut.validation:micronaut-validation"
-    implementation "io.micronaut:micronaut-runtime"
-    implementation "io.micronaut.grpc:micronaut-grpc-server-runtime"
-    implementation "io.micronaut.discovery:micronaut-discovery-client"
-    implementation "io.grpc:grpc-services"
-    implementation "io.grpc:grpc-protobuf"
-    implementation "com.google.protobuf:protobuf-java-util"
-    implementation "org.yaml:snakeyaml"
+    compileOnly libs.managed.grpc.stub
+    compileOnly libs.javax.annotation.api
+    compileOnly mnValidation.micronaut.validation.processor
 
-    runtimeOnly "ch.qos.logback:logback-classic"
+    implementation mnValidation.micronaut.validation
+    implementation mn.micronaut.runtime
+    implementation projects.micronautGrpcRuntime
+    implementation libs.micronaut.discovery.client
+    implementation libs.managed.grpc.services
+    implementation libs.managed.grpc.protobuf
+    implementation libs.managed.protobuf.java.util
+    implementation mn.snakeyaml
 
-    testCompileOnly "javax.annotation:javax.annotation-api"
+    runtimeOnly mnLogging.logback.classic
 
-    testImplementation "org.apache.groovy:groovy"
+    testCompileOnly libs.javax.annotation.api
 
-    testRuntimeOnly "ch.qos.logback:logback-classic"
+    testImplementation libs.groovy
+
+    testRuntimeOnly mnLogging.logback.classic
 }
 
 tasks.withType(GroovyCompile) {
@@ -57,7 +65,7 @@ sourceSets {
 }
 
 protobuf {
-    protoc { artifact = "com.google.protobuf:protoc:$protobufJavaVersion" }
+    protoc { artifact = "com.google.protobuf:protoc:$protobufVersion" }
     plugins {
         grpc { artifact = "io.grpc:protoc-gen-grpc-java:$grpcVersion" }
     }

--- a/test-suite-java/build.gradle
+++ b/test-suite-java/build.gradle
@@ -7,6 +7,13 @@ plugins {
     id "io.micronaut.build.internal.grpc-tests"
 }
 
+// tag::variables[]
+ext {
+    grpcVersion = libs.versions.managed.grpc.asProvider().get()
+    protobufVersion = libs.versions.managed.protobuf.asProvider().get()
+}
+// end::variables[]
+
 mainClassName = "helloworld.Application"
 micronaut {
     version.set(libs.versions.micronaut.platform.get())
@@ -21,23 +28,23 @@ micronaut {
 
 dependencies {
 
-    annotationProcessor "io.micronaut.validation:micronaut-validation-processor"
+    annotationProcessor mnValidation.micronaut.validation.processor
 
-    compileOnly "io.grpc:grpc-stub"
-    compileOnly "javax.annotation:javax.annotation-api"
+    compileOnly libs.managed.grpc.stub
+    compileOnly libs.javax.annotation.api
 
-    implementation "io.micronaut.validation:micronaut-validation"
-    implementation "io.micronaut:micronaut-runtime"
+    implementation mnValidation.micronaut.validation
+    implementation mn.micronaut.runtime
     implementation projects.micronautGrpcServerRuntime
-    implementation "io.micronaut.discovery:micronaut-discovery-client"
-    implementation "io.grpc:grpc-services"
-    implementation "org.yaml:snakeyaml"
+    implementation libs.micronaut.discovery.client
+    implementation libs.managed.grpc.services
+    implementation mn.snakeyaml
 
-    runtimeOnly "ch.qos.logback:logback-classic"
+    runtimeOnly mnLogging.logback.classic
 
-    testAnnotationProcessor "io.micronaut.validation:micronaut-validation-processor"
+    testAnnotationProcessor mnValidation.micronaut.validation.processor
 
-    testCompileOnly "javax.annotation:javax.annotation-api"
+    testCompileOnly libs.javax.annotation.api
 }
 
 // tag::config[]
@@ -52,7 +59,7 @@ sourceSets {
 }
 
 protobuf {
-    protoc { artifact = "com.google.protobuf:protoc:$protobufJavaVersion" }
+    protoc { artifact = "com.google.protobuf:protoc:$protobufVersion" }
     plugins {
         grpc { artifact = "io.grpc:protoc-gen-grpc-java:$grpcVersion" }
     }

--- a/test-suite-java/build.gradle
+++ b/test-suite-java/build.gradle
@@ -1,7 +1,7 @@
 plugins {
-    id 'java'
+    id "java"
     // tag::plugin[]
-    id "com.google.protobuf" version "$protobufGradleVersion"
+    alias(libs.plugins.protobuf)
     // end::plugin[]
     id "io.micronaut.application"
     id "io.micronaut.build.internal.grpc-tests"

--- a/test-suite-java/pom.xml
+++ b/test-suite-java/pom.xml
@@ -11,9 +11,9 @@
 
     <properties>
         <packaging>jar</packaging>
-        <micronaut.version>4.0.0</micronaut.version>
-        <grpc.version>1.55.1</grpc.version>
-        <protoc.version>3.25.0</protoc.version>
+        <micronaut.version>4.1.11</micronaut.version>
+        <grpc.version>1.59.1</grpc.version>
+        <protoc.version>3.25.1</protoc.version>
         <guava.version>32.1.3-jre</guava.version>
         <jdk.version>17</jdk.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
@@ -113,7 +113,7 @@
             <plugin>
                 <groupId>io.micronaut.build</groupId>
                 <artifactId>micronaut-maven-plugin</artifactId>
-                <version>3.5.4</version>
+                <version>4.1.0</version>
             </plugin>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>

--- a/test-suite-kotlin/build.gradle
+++ b/test-suite-kotlin/build.gradle
@@ -15,10 +15,11 @@ kotlin {
 
 // tag::variables[]
 ext {
-    grpcKotlinVersion = "1.4.0"
+    grpcVersion = libs.versions.managed.grpc.asProvider().get()
+    grpcKotlinVersion = libs.versions.managed.grpc.kotlin.get()
+    protobufVersion = libs.versions.managed.protobuf.asProvider().get()
 }
 // end::variables[]
-
 
 mainClassName = "helloworld.Application"
 micronaut {
@@ -33,27 +34,26 @@ micronaut {
 }
 
 dependencies {
-    kapt "io.micronaut.validation:micronaut-validation-processor"
-    implementation "io.micronaut.validation:micronaut-validation"
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
-    implementation "org.jetbrains.kotlin:kotlin-reflect"
-    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.7.3"
-    implementation "io.micronaut:micronaut-runtime"
+    kapt mnValidation.micronaut.validation.processor
+    implementation mnValidation.micronaut.validation
+    implementation libs.kotlin.stdlib.jdk8
+    implementation libs.kotlin.reflect
+    implementation libs.kotlinx.coroutines.core
+    implementation mn.micronaut.runtime
     implementation projects.micronautGrpcServerRuntime
-    implementation "org.yaml:snakeyaml"
-
+    implementation mn.snakeyaml
 
     // tag::dependencies[]
-    implementation "io.grpc:grpc-kotlin-stub:$grpcKotlinVersion"
-    implementation "io.grpc:grpc-services"
-    compileOnly "io.grpc:grpc-stub"
-    compileOnly "javax.annotation:javax.annotation-api"
+    implementation libs.managed.grpc.kotlin.stub
+    implementation libs.managed.grpc.services
+    compileOnly libs.managed.grpc.stub
+    compileOnly libs.javax.annotation.api
     // end::dependencies[]
 
-    runtimeOnly "ch.qos.logback:logback-classic"
-    runtimeOnly "com.fasterxml.jackson.module:jackson-module-kotlin"
+    runtimeOnly mnLogging.logback.classic
+    runtimeOnly libs.jackson.module.kotlin
 
-    testCompileOnly "javax.annotation:javax.annotation-api"
+    testCompileOnly libs.javax.annotation.api
 }
 
 // tag::config[]
@@ -69,9 +69,9 @@ sourceSets {
 }
 
 protobuf {
-    protoc { artifact = "com.google.protobuf:protoc:$protobufJavaVersion" }
+    protoc { artifact = "com.google.protobuf:protoc:$protobufVersion" }
     plugins {
-        grpc { artifact = "io.grpc:protoc-gen-grpc-java:${grpcVersion}" }
+        grpc { artifact = "io.grpc:protoc-gen-grpc-java:$grpcVersion" }
         grpckt { artifact = "io.grpc:protoc-gen-grpc-kotlin:${grpcKotlinVersion}:jdk8@jar" }
     }
     generateProtoTasks {

--- a/test-suite-kotlin/build.gradle
+++ b/test-suite-kotlin/build.gradle
@@ -3,7 +3,7 @@ plugins {
     id "org.jetbrains.kotlin.kapt"
     id "org.jetbrains.kotlin.plugin.allopen"
     id "io.micronaut.application"
-    id "com.google.protobuf" version "$protobufGradleVersion"
+    alias(libs.plugins.protobuf)
     id "io.micronaut.build.internal.grpc-tests"
 }
 

